File: src/main/java/supermetroid/memory/WindowsProcessReader.java
package supermetroid.memory;

import com.sun.jna.Memory;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;

public class WindowsProcessReader extends MemoryReader {
  private final Snes9x snes9x;
  private final int BYTE_READ_SIZE = 1;
  private final Memory BYTE_MEMORY = new Memory(BYTE_READ_SIZE);
  private final int USHORT_READ_SIZE = 2;
  private final Memory USHORT_MEMORY = new Memory(USHORT_READ_SIZE);
  private final IntByReference READ = new IntByReference();

  public WindowsProcessReader() {
    snes9x = Snes9x.get();
  }

  @Override
  public byte readByte(int offset) {
    Snes9x.kernel32.ReadProcessMemory(snes9x.handle, Pointer.createConstant(snes9x.baseAddress + offset), BYTE_MEMORY, BYTE_READ_SIZE, READ);
    return BYTE_MEMORY.getByte(0);
  }

  @Override
  public int readShort(int offset) {
    Snes9x.kernel32.ReadProcessMemory(snes9x.handle, Pointer.createConstant(snes9x.baseAddress + offset), USHORT_MEMORY, USHORT_READ_SIZE, READ);
    return Short.toUnsignedInt(USHORT_MEMORY.getShort(0));
  }
}

File: src/main/java/supermetroid/memory/WindowsProcessReader.kt
package supermetroid.memory

import com.sun.jna.Memory
import com.sun.jna.Pointer
import com.sun.jna.ptr.IntByReference

class WindowsProcessReader : MemoryReader() {
    private val snes9x: Snes9x = Snes9x.get()
    private val read = IntByReference()

    private val byteSize = 1
    private val byteMemory = Memory(byteSize.toLong())
    override fun readByte(offset: Int): Int {
        Snes9x.kernel32.ReadProcessMemory(
            snes9x.handle,
            Pointer.createConstant(snes9x.baseAddress + offset),
            byteMemory,
            byteSize,
            read
        )
        return byteMemory.getByte(0).toInt() and 255
    }

    private val shortSize = 2
    private val shortMemory = Memory(shortSize.toLong())
    override fun readShort(offset: Int): Int {
        Snes9x.kernel32.ReadProcessMemory(
            snes9x.handle,
            Pointer.createConstant(snes9x.baseAddress + offset),
            shortMemory,
            shortSize,
            read
        )
        return java.lang.Short.toUnsignedInt(shortMemory.getShort(0))
    }
}
